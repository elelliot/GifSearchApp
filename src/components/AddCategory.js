import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory =  ( {setCategories} ) => {

    const [inputValue, setInputValue] = useState('');//string vacío, si lo dejamos sin '', es undefined y causará errores

    const handleInputChange = ( e ) => {
        //console.log(e.target.value);
        setInputValue (e.target.value);//obtenemos los cambios de texto con el evento

        //console.log('handleInputChange llamado');
    }

    const handleSubmit = ( e ) =>{

        e.preventDefault()//prevenimos el comportamiento por defecto que sería refrescar
        // console.log('Submit Hecho')

        //console.log('handleSubmit', inputValue);


        //validamos que no haya valores nulos o que sean mínimo de 3 caracteres
        if( inputValue.trim().length > 2){
            
            /*como no pasé las categories como argumento junto con el setCategories 
            llamo al callback que tiene el estado anterior, en este caso le llamo cats*/
            setCategories( cats => [ inputValue, ...cats]);//insertamos primero la categoría que buscamos y despues las anteriores
            setInputValue('');
        }


    }


    return (
        //cuando se quiere regresar un formulario, los fragments son redundantes ya que el form agrupa o el div
        //cuando aplastas enter el form se envía y se refresca por defecto
        <form onSubmit = { handleSubmit }>
            <p> { inputValue }</p>
            <input 
                type="text"
            /*el value se queda igual ya que esto se redibuja cada que toco el teclado y agarramos el valor desde el evento onChange*/
               value= { inputValue }
            /*Se puede mandar el primer argumento del onChange que sería el evento y lo atrapamos en la función (arriba) */
                onChange = { handleInputChange /*e => console.log(e)*/ } 
            />
        </form>
        
    )
}
//hacemos que el setCategories sea funcion obligatoria
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
