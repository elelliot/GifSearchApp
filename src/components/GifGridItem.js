import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ( { /*id,*/ title, url } ) => {

    // console.log( id, title, url  );

    return (
        <div className="card animate__animated animate__fadeIn ">
            <img src= { url } alt = { title } />
            <p>{ title }</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}
// import React from 'react'

// export const GifGridItem = ( img ) => {

//     console.log( img );
    
//     return (
//         <div>
//             { img.title }
//         </div>
//     )
// }

















