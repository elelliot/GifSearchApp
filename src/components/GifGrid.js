import React from 'react';
import Proptypes from 'prop-types';

import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //usamos el custom hook
    const { data:images , loading } = useFetchGifs( category ); //renombramos la data desestructurada


    // const [images, setImages] = useState([])


    return (
        <>
            <h3 className=" animate__animated animate__fadeIn ">{ category }</h3>

            {/* para no tener que poner un null si la condicion no se cumple usamos el && para solo evaluar lo primero 
            y no hacer nada si no se cumple */}
            { loading && <p className = "animate__animated animate__flash">Loading</p> }
            {/*  loading ? <p>Loading</p> : null */}
            
            <div className = "card-grid">
                
                {
                    images.map( img => (
                        <GifGridItem 
                            key = { img.id }
                            { ...img } 
                            // img = { img } 
                        />
                        // return <li key= { imagen.id }>{ imagen.title }</li>
                    ))
                    
                }
                
            </div>
        </>
    )
}


GifGrid.propTypes = {
    category: Proptypes.string.isRequired
}

