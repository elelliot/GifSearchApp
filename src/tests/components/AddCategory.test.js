import { shallow } from "enzyme"
import { AddCategory } from "../../components/AddCategory"
import '@testing-library/jest-dom'

describe('Pruebas en <AddCategory />', () => {

    //const setCategories = () => {}; //es una función vacía solo para llenar el argumento
    const setCategories = jest.fn();// lo mismo que arriba pero ahora se puede saber como , si fue, o cuantas veces fue llamada
    
    //se puede inicializar con ; para no repetir, pero se hace para acceder a los metodos
    let wrapper = shallow( <AddCategory setCategories = { setCategories } /> );
    


    
    beforeEach( () => {
        jest.clearAllMocks();//se llama para limpiar todo en una simulacion
        wrapper = shallow( <AddCategory setCategories = { setCategories } /> );
    })




    test('Debe mostrar el componente correctamente', () => {
    
        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe cambiar la caja de texto', () => {
        
        const input = wrapper.find('input');

        const value = 'Hola Demonio';

        //se manda el evento de un input change, este se define con {} , ese objeto tiene el valor del evento
        //recordar que existe el valor e.target.value el cual es el valor de lo que escribimos
        input.simulate('change', { target: { value } } );

        expect( wrapper.find('p').text().trim() ).toBe( value ); //esperamos que el parrafo = value
    })
    

    test('No debe postear la información con Submit', () => {
        
        //simulamos el submit del form
        wrapper.find('form').simulate('submit', { preventDefault(){} }); //el argumento del evento se manda solo, pero hay que mandar la función
        
        //como setCategories se llama gracias al Hola demonio de la prueba pasada, este debe reiniciarse con el beforeEach
        expect( setCategories ).not.toHaveBeenCalled(); 


    })


    test('Debe llamar el setCategories y limpiar la caja de texto', () => {
        
        
        const value = 'Se Agrego algo al input';

        const input = wrapper.find('input');
        //1. simular el inputChange
        input.simulate('change', { target: { value } } );

        //2. simular el submit
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        //3. setCategories  se debe de haber llamado
        expect( setCategories ).toHaveBeenCalled(); 
        // expect( setCategories ).toHaveBeenCalledTimes(1); 
        // expect( setCategories ).toHaveBeenCalledWith( expect.any(Function)); 

        //4. El valor del input debe ser '' (String vacío) 
        expect( input.prop('value') ).toBe( '' );




    })
    
    

    
})
