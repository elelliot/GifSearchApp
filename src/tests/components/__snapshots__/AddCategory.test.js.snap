// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Debe mostrar el componente correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
     
  </p>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
